server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java

```diff
@@ -354,4 +354,32 @@ public void removeGroupMappingShouldDelete() throws ErrorMappingException, Recip
         assertThat(virtualUserTable.getMappings(user, domain))
             .isEqualTo(MappingsImpl.empty());
     }
+
+    @Test
+    public void listSourcesShouldReturnWhenHasMapping() throws RecipientRewriteTableException {
+        String user = "test";
+        Domain domain = Domain.LOCALHOST;
+        String address = "test@localhost2";
+        MappingSource source = MappingSource.fromUser(user, domain);
+        Mapping mapping = Mapping.group(address);
+        virtualUserTable.addMapping(source, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source);
+    }
+
+    @Test
+    public void listSourceShouldReturnWhenMultipleSourceMapping() throws RecipientRewriteTableException {
+        String user = "test";
+        Domain domain = Domain.of("james");
+        String address = "test@localhost2";
+
+        MappingSource source = MappingSource.fromUser(user, domain);
+        MappingSource source2 = MappingSource.fromDomain(Domain.LOCALHOST);
+        Mapping mapping = Mapping.group(address);
+
+        virtualUserTable.addMapping(source, mapping);
+        virtualUserTable.addMapping(source2, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source, source2);
+    }
 }
```

Comment:
You are missing...

 - Tests with forward + address that should be handled...
 - Tests that a mapping without source will lead to an empty list being returned.

Reply:
Add more tests